File: tsbinding/opencv/Algorithm.ts
////// <reference path="Matrix.ts" />
var alvision_module = require('../../lib/bindings.js');

import * as _constants from './Constants'
import * as _st from './static'


//module alvision {
export interface Algorithm {
    clear(): void;
    //write(fs: FileStorage): void;
    empty(): boolean;
}

export interface AlgorithmStatic {
    new (): Algorithm;
}

export var Algorithm: AlgorithmStatic = alvision_module.Algorithm;



///** @brief Reads algorithm parameters from a file storage
//*/
//virtual void read(const FileNode& fn) { (void)fn; }

///** @brief Returns true if the Algorithm is empty (e.g. in the very beginning or after unsuccessful read
// */
//virtual bool empty() const { return false; }

///** @brief Reads algorithm from the file node

// This is static template method of Algorithm. It's usage is following (in the case of SVM):
// @code
// Ptr<SVM> svm = Algorithm::read<SVM>(fn);
// @endcode
// In order to make this method work, the derived class must overwrite Algorithm::read(const
// FileNode& fn) and also have static create() method without parameters
// (or with all the optional parameters)
// */
//template < typename _Tp> static Ptr< _Tp > read(const FileNode& fn)
//    {
//        Ptr<_Tp>obj = _Tp::create();
//obj ->read(fn);
//return !obj ->empty() ? obj : Ptr<_Tp>();
//    }

///** @brief Loads algorithm from the file

// @param filename Name of the file to read.
// @param objname The optional name of the node to read (if empty, the first top-level node will be used)

// This is static template method of Algorithm. It's usage is following (in the case of SVM):
// @code
// Ptr<SVM> svm = Algorithm::load<SVM>("my_svm_model.xml");
// @endcode
// In order to make this method work, the derived class must overwrite Algorithm::read(const
// FileNode& fn).
// */
//template < typename _Tp> static Ptr< _Tp > load(const String& filename, const String& objname=String())
//    {
//        FileStorage fs(filename, FileStorage::READ);
//FileNode fn = objname.empty() ? fs.getFirstTopLevelNode() : fs[objname];
//Ptr < _Tp > obj = _Tp::create();
//obj ->read(fn);
//return !obj ->empty() ? obj : Ptr<_Tp>();
//    }

///** @brief Loads algorithm from a String

// @param strModel The string variable containing the model you want to load.
// @param objname The optional name of the node to read (if empty, the first top-level node will be used)

// This is static template method of Algorithm. It's usage is following (in the case of SVM):
// @code
// Ptr<SVM> svm = Algorithm::loadFromString<SVM>(myStringModel);
// @endcode
// */
//template < typename _Tp> static Ptr< _Tp > loadFromString(const String& strModel, const String& objname=String())
//    {
//        FileStorage fs(strModel, FileStorage::READ + FileStorage::MEMORY);
//FileNode fn = objname.empty() ? fs.getFirstTopLevelNode() : fs[objname];
//Ptr < _Tp > obj = _Tp::create();
//obj ->read(fn);
//return !obj ->empty() ? obj : Ptr<_Tp>();
//    }

///** Saves the algorithm to a file.
// In order to make this method work, the derived class must implement Algorithm::write(FileStorage& fs). */
//CV_WRAP virtual void save(const String& filename) const;

///** Returns the algorithm string identifier.
// This string is used as top level xml/yml node tag when the object is saved to a file or string. */
//CV_WRAP virtual String getDefaultName() const;
//};

//struct Param {
//    enum { INT = 0, BOOLEAN = 1, REAL = 2, STRING = 3, MAT = 4, MAT_VECTOR = 5, ALGORITHM = 6, FLOAT = 7,
//        UNSIGNED_INT = 8, UINT64 = 9, UCHAR = 11
//    };
//};

File: src/opencv/core/Algorithm.cc
#include "Algorithm.h"
#include "../persistence/FileNode.h"
#include "../persistence/FileStorage.h"

namespace algorithm_general_callback {
	std::shared_ptr<overload_resolution> overload;
	NAN_METHOD(callback) {
		if (overload == nullptr) {
			throw std::exception("algorithm_general_callback is empty");
		}
		return overload->execute("algorithm", info);
	}
}

Nan::Persistent<FunctionTemplate> Algorithm::constructor;


void
Algorithm::Init(Handle<Object> target, std::shared_ptr<overload_resolution> overload) {


	//Class
	Local<FunctionTemplate> ctor = Nan::New<FunctionTemplate>(algorithm_general_callback::callback);
	constructor.Reset(ctor);
	ctor->InstanceTemplate()->SetInternalFieldCount(1);
	ctor->SetClassName(Nan::New("Algorithm").ToLocalChecked());

	overload->register_type<Algorithm>(ctor, "algorithm", "Algorithm");


	overload->addStaticOverload("algorithm", "Algorithm", "read", {
			make_param<std::string>("Ttype","String"),
			make_param<FileNode*>("fn","FileNode")
	}, read);
	//read<T>(Ttype: string, fn : _persistence.FileNode) : T;

	//template < typename _Tp> static Ptr< _Tp > read(const FileNode& fn)
	//    {
	//        Ptr<_Tp>obj = _Tp::create();
	//obj ->read(fn);
	//return !obj ->empty() ? obj : Ptr<_Tp>();
	//    }


	overload->addStaticOverload("algorithm", "Algorithm", "load", {
		   make_param<std::string>("Ttype","String"),
		   make_param<std::string>("filename","String"),
		   make_param<std::string>("objname","String")
	}, load);
	//load<T>(Ttype : string, filename : string, objname ? : string) : T;
	//template < typename _Tp> static Ptr< _Tp > load(const String& filename, const String& objname=String())
	//    {
	//        FileStorage fs(filename, FileStorage::READ);
	//FileNode fn = objname.empty() ? fs.getFirstTopLevelNode() : fs[objname];
	//Ptr < _Tp > obj = _Tp::create();
	//obj ->read(fn);
	//return !obj ->empty() ? obj : Ptr<_Tp>();
	//    }



	overload->addStaticOverload("algorithm", "Algorithm", "loadFromString", {
		make_param<std::string>("Ttype","String"),
		make_param<std::string>("strModel","String"),
		make_param<std::string>("objname","String")
	}, loadFromString);
	//loadFromString<T>(Ttype : string, strModel : string, objname ? : string) : T;
	//template < typename _Tp> static Ptr< _Tp > loadFromString(const String& strModel, const String& objname=String())
	//    {
	//        FileStorage fs(strModel, FileStorage::READ + FileStorage::MEMORY);
	//FileNode fn = objname.empty() ? fs.getFirstTopLevelNode() : fs[objname];
	//Ptr < _Tp > obj = _Tp::create();
	//obj ->read(fn);
	//return !obj ->empty() ? obj : Ptr<_Tp>();
	//    }
//}


	//clear() : void;
	overload->addOverload("algorithm", "Algorithm", "clear", {}, clear);


	overload->addOverload("algorithm", "Algorithm", "write", {
		make_param<FileStorage*>("fs","FileStorage")
	}, write);
	//write(fs: _persistence.FileStorage) : void;



	overload->addOverload("algorithm", "Algorithm", "read", {
		make_param<FileNode*>("fn","FileNode")
	}, read_member);
	//read(fn: _persistence.FileNode) : void;


	overload->addOverload("algorithm", "Algorithm", "empty", {}, empty);
	//empty() : boolean;




	overload->addOverload("algorithm", "Algorithm", "save", {
		make_param<std::string>("filename","String")
	}, save);
	//save(filename : string) : void;


	overload->addOverload("algorithm", "Algorithm", "getDefaultName", {}, getDefaultName);
	//getDefaultName() : string;

//struct Param {
//    enum { INT = 0, BOOLEAN = 1, REAL = 2, STRING = 3, MAT = 4, MAT_VECTOR = 5, ALGORITHM = 6, FLOAT = 7,
//        UNSIGNED_INT = 8, UINT64 = 9, UCHAR = 11
//    };
//};




	target->Set(Nan::New("Algorithm").ToLocalChecked(), ctor->GetFunction());


};

v8::Local<v8::Function> Algorithm::get_constructor() {
	assert(!constructor.IsEmpty() && "constructor is empty");
	return Nan::New(constructor)->GetFunction();
}

POLY_METHOD(Algorithm::New) {
	auto algorithm = new Algorithm();

	algorithm->Wrap(info.Holder());
	info.GetReturnValue().Set(info.Holder());
}

POLY_METHOD(Algorithm::read){
	throw std::exception("not implemented");
}
POLY_METHOD(Algorithm::load){
	throw std::exception("not implemented");
}
POLY_METHOD(Algorithm::loadFromString){
	throw std::exception("not implemented");
}
POLY_METHOD(Algorithm::clear){
	throw std::exception("not implemented");
}
POLY_METHOD(Algorithm::write){
	throw std::exception("not implemented");
}
POLY_METHOD(Algorithm::read_member){
	throw std::exception("not implemented");
}
POLY_METHOD(Algorithm::empty){
	throw std::exception("not implemented");
}
POLY_METHOD(Algorithm::save){
	throw std::exception("not implemented");
}
POLY_METHOD(Algorithm::getDefaultName){
	throw std::exception("not implemented");
}